package/utils/px5g-wolfssl/px5g-wolfssl.c

```diff
@@ -0,0 +1,312 @@
+      days = (unsigned int)atoi(arg[1]);
+      arg++;
+    } else if (!strncmp(*arg, "-pkeyopt", 8) && arg[1]) {
+      if (strncmp(arg[1], "ec_paramgen_curve:", 18)) {
+        fprintf(stderr, "error: invalid pkey option: %s\n", arg[1]);
+        return 1;
+      }
+      if (!strncmp(arg[1] + 18, "P-256:", 5)) {
+        curve = ECC_SECP256R1;
+      } else if (!strncmp(arg[1] + 18, "P-384:", 5)) {
+        curve = ECC_SECP384R1;
+      } else if (!strncmp(arg[1] + 18, "P-521:", 5)) {
+        curve = ECC_SECP521R1;
+      } else {
+        fprintf(stderr, "error: invalid curve name: %s\n", arg[1] + 18);
+        return 1;
+      }
+      arg++;
+    } else if (!strncmp(*arg, "-keyout", 7) && arg[1]) {
+      keypath = arg[1];
+      arg++;
+    } else if (!strncmp(*arg, "-out", 4) && arg[1]) {
+      certpath = arg[1];
+      arg++;
+    }
+    arg++;
+  }
+
+  gen_key(rng, &ecKey, &rsaKey, type, keySz, exp, curve);
+  write_key(&ecKey, &rsaKey, type, keySz, keypath, pem);
+
+  from = (from < 1000000000) ? 1000000000 : from;
+  strftime(fstr, sizeof(fstr), "%Y%m%d%H%M%S", gmtime(&from));
+  to = from + 60 * 60 * 24 * days;
+  if (to < from)
+    to = INT_MAX;
+  strftime(tstr, sizeof(tstr), "%Y%m%d%H%M%S", gmtime(&to));
+
+  fprintf(stderr,
+          "Generating selfsigned certificate with subject '%s'"
+          " and validity %s-%s\n",
+          subject, fstr, tstr);
+
+  memset(pemBuf, 0, sizeof(pemBuf));
+  ret = wc_InitCert(&newCert);
+  if (ret != 0) {
+    fprintf(stderr, "Init Cert failed: %d\n", ret);
+    return ret;
+  }
+  newCert.isCA = 0;
+
+  if (type == EC_KEY_TYPE) {
+    ret = wc_MakeCert(&newCert, derBuf, sizeof(derBuf), NULL, &ecKey, rng);
+  } else {
+    ret = wc_MakeCert(&newCert, derBuf, sizeof(derBuf), &rsaKey, NULL, rng);
+  }
+  if (ret <= 0) {
+    fprintf(stderr, "Make Cert failed: %d\n", ret);
+    return ret;
+  }
+
+  if (type == EC_KEY_TYPE) {
+    newCert.sigType = CTC_SHA256wECDSA;
+    ret = wc_SignCert(newCert.bodySz, newCert.sigType, derBuf, sizeof(derBuf),
+                      NULL, &ecKey, rng);
+  } else {
+    newCert.sigType = CTC_SHA256wRSA;
+    ret = wc_SignCert(newCert.bodySz, newCert.sigType, derBuf, sizeof(derBuf),
+                      &rsaKey, NULL, rng);
+  }
+  if (ret <= 0) {
+    fprintf(stderr, "Sign Cert failed: %d\n", ret);
+    return ret;
+  }
+  derSz = ret;
+
+  ret = wc_DerToPem(derBuf, derSz, pemBuf, sizeof(pemBuf), CERT_TYPE);
+  if (ret <= 0) {
+    fprintf(stderr, "DER to PEM failed: %d\n", ret);
+    return ret;
+  }
+  pemSz = ret;
+
+  ret = write_file(pemBuf, pemSz, certpath);
+  if (ret != 0) {
+    fprintf(stderr, "Write Cert failed: %d\n", ret);
+    return ret;
+  }
+
+  if (type == EC_KEY_TYPE) {
+    wc_ecc_free(&ecKey);
+  } else {
+    wc_FreeRsaKey(&rsaKey);
+  }
+  return 0;
+}
+
+int dokey(WC_RNG *rng, int type, char **arg) {
+  ecc_key ecKey;
+  RsaKey rsaKey;
+  int ret;
+  int curve = ECC_SECP256R1;
+  int keySz = 2048;
```

Comment:
*maybe* change to WOLFSSL_MIN_RSA_BITS . It's kind of pushing it.

Reply:
Shouldn't we use 4096 when we already use a validity of 10 years?

Reply:
No. Not only is 4096 slow, it doesn't buy anything. 2048 will not be broken in 10 years.

The difference between RSA and EC is vulnerability to index calculus. If that were not the case, RSA256 == EC256. But since it is not, RSA needs to use ridiculous key lengths. EC256 == RSA3072 ignoring implementation problems and whatever else.

Actually now that I think about it, using WOLFSSL_MIN_RSA_BITS is perfect. In the unlikely even that RSA 2048 becomes insecure, px5g would be automatically migrated.

Reply:
Fine for me. Are you okay with Ecdsa as default?

Reply:
I am. Although I don't maintain uhttpd.

Reply:
WOLFSSL_MIN_RSA_BITS doesn't seem to be available, it's only defined in wolfssl/internal.h.

Reply:
Ah I missed that.